packages/protobuf/src/next/from-binary.ts

```diff
@@ -0,0 +1,274 @@
+} from "../google/protobuf/descriptor_pb.js";
+
+// Default options for parsing binary data.
+const readDefaults: Readonly<BinaryReadOptions> = {
+  readUnknownFields: true,
+  readerFactory: (bytes) => new BinaryReader(bytes),
+};
+
+function makeReadOptions(
+  options?: Partial<BinaryReadOptions>,
+): Readonly<BinaryReadOptions> {
+  return options ? { ...readDefaults, ...options } : readDefaults;
+}
+
+export function fromBinary<Desc extends DescMessage>(
+  desc: Desc,
+  bytes: Uint8Array,
+  options?: Partial<BinaryReadOptions>,
+): MessageShape<Desc> {
+  options = makeReadOptions(options);
+  return readMessage(
+    desc,
+    options.readerFactory!(bytes),
+    bytes.byteLength,
+    options as BinaryReadOptions,
+  );
+}
+
+function readMessage<Desc extends DescMessage>(
+  desc: Desc,
+  reader: IBinaryReader,
+  lengthOrEndTagFieldNo: number,
+  options: BinaryReadOptions,
+  delimitedMessageEncoding?: boolean,
+): MessageShape<Desc> {
+  const message = create(desc);
+  const rMessage = reflect(message);
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  const end = delimitedMessageEncoding
+    ? reader.len
+    : reader.pos + lengthOrEndTagFieldNo;
+  let fieldNo: number | undefined, wireType: WireType | undefined;
+  while (reader.pos < end) {
+    [fieldNo, wireType] = reader.tag();
+    if (wireType == WireType.EndGroup) {
+      break;
+    }
+    const field = rMessage.findNumber(fieldNo);
+    if (!field) {
+      reader.skip(wireType);
+      // TODO: read unknown fields
+      continue;
+    }
+    readField(rMessage, reader, field, wireType, options);
+  }
+  if (
+    delimitedMessageEncoding && // eslint-disable-line @typescript-eslint/strict-boolean-expressions
+    (wireType != WireType.EndGroup || fieldNo !== lengthOrEndTagFieldNo)
+  ) {
+    throw new Error(`invalid end group tag`);
+  }
+  return message;
+}
+
+function readField(
+  message: ReflectMessage,
+  reader: IBinaryReader,
+  field: DescField,
+  wireType: WireType,
+  options: BinaryReadOptions,
+) {
+  switch (field.fieldKind) {
+    case "scalar":
+      message.set(field, readScalar(reader, field.scalar, field.longType));
+      break;
+    case "enum":
+      message.set(field, readScalar(reader, ScalarType.INT32) as number);
+      break;
+    case "message":
+      message.set(field, readMessageField(reader, options, field));
+      break;
+    case "list":
+      readRepeatedField(message, reader, options, field, wireType);
+      break;
+    case "map":
+      // eslint-disable-next-line no-case-declarations
+      const [key, val] = readMapEntry(field, reader, options);
+      message.setMapEntry(field, key, val);
+      break;
+  }
+}
+
+// Read a map field, expecting key field = 1, value field = 2
+function readMapEntry(
+  field: DescField & { fieldKind: "map" },
+  reader: IBinaryReader,
+  options: BinaryReadOptions,
+): [string | number, ScalarValue | MessageShape<DescMessage>] {
```

Comment:
We have to update this logic to support edition features. They apply to synthetic map entries as well. 

Can you add a TODO?